core/src/main/java/org/infinispan/commands/write/RemoveExpiredCommand.java

```diff
@@ -85,15 +93,27 @@ public Object perform(InvocationContext ctx) throws Throwable {
          } else if (e.getLifespan() > 0 && e.getLifespan() == lifespan) {
             // If the entries lifespan is not positive that means it can't expire so don't even try to remove it
             // Lastly if there is metadata we have to verify it equals our lifespan and the value match.
-            // TODO: add a threshold to verify this wasn't just created with the same value/lifespan just before expiring
             if (valueMatcher.matches(prevValue, value, null)) {
-               e.setExpired(true);
-               return performRemove(e, prevValue, ctx);
+               // Only remove the entry if it is still expired.
+               // Due to difference in system clocks between nodes - we also accept the expiration within 100 ms
+               // of now. This along with the fact that entries are not normally access immediately when they
+               // expire should give a good enough buffer range to not create a false positive.
```

Comment:
@wburns I'm not sure why we need this, didn't we already check that the entry is expired on all the owners before invoking `performExpire()`?

Reply:
We do, but this is usually done outside of a lock such as a get. In this case we have to acquire the lock during remove expired and double check it is still expired (in case if a concurrent write happened).

Reply:
Right, but do we even acquire the lock if the entry was not expired on all the owners during the get?

Reply:
hrmm, there is no lock acquisition at all during a get. We don't fire the remove expired command during a get if the entry wasn't expired. Or maybe you are trying to ask something else?